### DISCUSSION TITLE
Selfhosted Mastodon can't follow self hosted Pixelfed on same server

### DISCUSSION THREAD

**Discussion by chrisale** [mastodon/mastodon#28501] [General]
### Steps to reproduce the problem

1. Install Mastodon on Debian 11 on a server on a home NAT based network.
2. Install Pixelfed on same Debian 11 (in separate linux user)
3. From the Mastodon web client interface, attempt to search or follow local Pixelfed account @person@pixelfed.mysame.server from local Mastodon account @person@mastodon.mysame.server
4. Search returns '503Remote data could not be fetched'.

### Expected behaviour

Mastodon user should follow Pixelfed users

### Actual behaviour

Failed: 503Remote data could not be fetched

### Detailed description

I have also created a ticket for this on the Pixelfed GitHub at:
https://github.com/pixelfed/pixelfed/issues/4657

I have Pixelfed and Mastodon running natively together, in separate user accounts, on a Debian 11 server. All appears to be working, people can search for, discover and follow @[chris@pixelfed.chrisalemany.ca](mailto:chris@pixelfed.chrisalemany.ca) from any software on the Fediverse and I can interact normally from Pixelfed. Likewise, my Mastodon account is fully functional.

The exception is the connection between my local Pixelfed and Mastodon accounts.

When I search for @[chris@pixelfed.chrisalemany.ca](mailto:chris@pixelfed.chrisalemany.ca) from the mastodon web client at mstdn.chrisalemany.ca, Mastodon fails and reports:
503Remote data could not be fetched

nginx/access.log logs the following:

[18/Sep/2023:08:35:59 -0700] "GET /api/v2/search?q=%40chris%40pixelfed.chrisalemany.ca&resolve=true&limit=5 HTTP/2.0" 503 44 "https://mstdn.chrisalemany.ca/getting-started" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Safari/605.1.15" 
Nothing is produced in the nginx/error log.

I am running Mastodon 4.1.7 and Pixelfed 0.11.9
Enabling ElasticSearch on Mastodon has no effect. (I generally don't have it enabled)

The second issue is likely related to the first, so I will only put it up here for added context. I can see and follow my mastodon account @[chris@mstdn.chrisalemany](mailto:chris@mstdn.chrisalemany) from @[chris@pixelfed.chrisalemany.ca](mailto:chris@pixelfed.chrisalemany.ca). However, new posts from my Mastodon do not actually appear in my home Pixelfed timeline. Again, I suspect this is just another symptom of the same issue as causes the first problem with following from Mastodon.

### Mastodon instance

mstdn.chrisalemany.ca

### Mastodon version

v4.1.7

### Technical details

Selfhosted: Following official documented procedures for Mastodon and Pixelfed installs

ruby 3.0.6p216 (2023-03-30 revision 23a532679b) [x86_64-linux]
node version: v16.20.2
nginx version: nginx/1.18.0

Server is running a local DNS so clients on the home NAT can connect to the server without needing loopback or other services not provided by the ISP.

**Comment by nightpool** [comment]
What is the actual result of resolving `pixelfed.mysame.server` as a DNS name from the instance in question? do you have that name in your /etc/hosts file by any chance? If so, this is not supported setup since it is a type of server-side request forgery—Mastodon will only interact with instances hosted on publicly routable IP addresses to avoid the possibility of Mastodon being used to compromise an internal network service (e.g. elasticsearch bound to the loopback interface)

**Comment by chrisale** [comment]
The result is an outside address (pixelfed.chrisalemany.ca). However, as the server is on a local NAT without access to loopback or local DNS from my ISP, I am running a DNS on the Mastodon/Pixelfed server and that includes having the pixelfed.chrisalemany.ca and mstdn.chrisalemany.ca in the /etc/hosts file so that clients can see the server on my local network.

It makes perfect sense that this would be the problem. The question is, I guess, is there a solution for self hosted single/small user instances like mine who can only run their machines behind a NAT?

Thanks for the insight.

**Comment by chrisale** [comment]
I've removed pixelfed.chrisalemany.ca from /etc/hosts for now so that only the mastodon address is specified there.  The remote data error continues. Are there any other strategies I could employ to get work with the security implications here?

**Comment by b2cc** [comment]
We're running a similar setup, with multiple Fediverse services behind a single NAT. You most likely will not be able to resolve this with local DNS alone. You will at least need a reverse proxy, which can forward traffic between the services and is also able to distinguish traffic by HTTP server name headers, or TLS SNI to correctly pass traffic between applications.

We resolved this by using pfsense as a router, with a 1:1 NAT and port-forward for EXTERNAL traffic to the reverse proxy on the application server that runs the applications.

Additionally we also run a haproxy on the pfsense router listening on the INTERNAL IP interface, that passes back traffic from one application to another. E.g. traffic from Mastodon to Pixelfed leaves the server, hits the haproxy on the router, and gets passed back to the application server where it came from, and then gets accepted by Pixelfed, avoiding the HTTP/503 as it now looks like "real" traffic entering the server from the outside.

Specifically for pfsense (or any other router with a webinterface on ports 80/443) also check this comment to not accidentally intercept outgoing requests on the router webinterface:
* https://github.com/cert-manager/cert-manager/issues/2517#issuecomment-889961035


It might be possible to resolve this locally with some clever local NATting/routing and by using a sophisticated reverse proxy setup as well, but that's just an educated guess from my side.
